README
Overview
========
Hopscotch is a framework to make it easy for developers to add product tours to their pages. Hopscotch accepts a tour JSON object as input and provides an API for the developer to control rendering the tour display and managing the tour progress.

Features
========
* Callbacks for tour events (e.g. onStart, onEnd, onShow, onNext, onPrev, onClose)
* Multi-page persistence using HTML5 sessionStorage using cookies as a fallback
* I18N support

General Usage
=============

To get started using the Hopscotch framework, simply include hopscotch.js on your page. This will load the hopscotch object into the global window object for you. Also be sure to include hopscotch.css to get the default styling.

    <html>
      <head>
        <title>My First Hopscotch Tour</title>
        <link rel="stylesheet" href="css/hopscotch.css"></link>
      </head>
      <body>
        <div>
          <p>Content goes here...</p>
        </div>
        <script src="js/hopscotch.js"></script>
      </body>
    </html>

Defining a Tour
===============
A Hopscotch tour consists of a tour id, an array of tour steps defined as JSON objects, and a number of tour-specific options. The tour id is simply a unique identifier string. The simplest tour consists of just an id string and an array of one or more steps.

Basic step options
------------------
The step options below are the most basic options.

    {
      id: {STRING - id of the tour},
      steps: [
        {
          // MANDATORY
          target:         STRING/ELEMENT - id of the target DOM element or DOM element itself,
          orientation:    STRING         - ["top", "bottom", "right", "left"]
          // OPTIONAL
          title:          STRING         - step title,
          content:        STRING         - step content
        },
        ...
      ]
    };

Note that title and content are both optional only because you can choose to have a step with only a title, only content, or both title and content.

This is an example of a tour defined with only basic steps.

    {
      id: "welcome_tour",
      steps: [
        {
          target: "header",
          orientation: "bottom",
          title: "This is the navigation menu",
          content: "Use the links here to get around on our site!"
        },
        {
          target: "profile-pic",
          orientation: "right",
          title: "Your profile picture",
          content: "Upload a profile picture here. This is the image that others will see next to your activity."
        },
        {
          target: "inbox",
          orientation: "bottom",
          title: "Your inbox",
          content: "Messages from other users will appear here."
        }
      ]
    }

**IMPORTANT** -- title and content are set using element.innerHTML. This allows the inclusion of very basic markup like links and lists. However, it also allows the inclusion of malicious script injections when used improperly. It is highly recommended to never show user-generated content in a Hopscotch tour. If it is absolutely necessary, you must properly escape the input, as always.

All step options
----------------
The comprehensive list of step options are listed below:

### MANDATORY ###

**target** [STRING/ELEMENT/ARRAY] - id of the target DOM element or DOM element itself. It is also possible to define an array of several targets. If an array is provided, Hopscotch will use the first target that exists on the page and disregard the rest.

**orientation** [STRING] - specifies where the bubble should appear in relation to the target. Valid values are "top", "bottom", "right", "left".

### OPTIONAL ###

**title** [STRING] - step title

**content** [STRING] - step content

**width** [INT] - bubble width

**padding** [INT] - bubble padding

**xOffset** [INT] - horizontal position adjustment for bubble. *Default*: 0.

**yOffset** [INT] - vertical position adjustment for bubble. *Default*: 0.

**arrowOffset** [INT] - offset for the bubble arrow. *Default*: 0.

**delay** [INT] - number in milliseconds to wait before showing step. *Default*: 0.

**zindex** [INT] - sets the z-index of the bubble

**showNextButton** [BOOLEAN] - should show the next button. *Default*: true.

**showPrevButton** [BOOLEAN] - should show the prev button. *Default*: true.

**showCTAButton** [BOOLEAN] - should show the call-to-action button. *Default*: false.

**ctaLabel** [STRING] - label for the call-to-action button.

**multipage** [BOOLEAN] - indicates that the next step is on a different page. *Default*: false.

**showSkip** [BOOLEAN] - if true, 'Next' button reads 'Skip'. *Default*: false.

**fixedElement** [BOOLEAN] - set to true if the target element has fixed positioning. *Default*: false.

**nextOnTargetClick** [BOOLEAN] - triggers nextStep() when the target is clicked. *Default*: false.

**onPrev** [FUNCTION] - callback for when 'Previous' button is clicked}

**onNext** [FUNCTION] - callback for when 'Next' button is clicked}

**onShow** [FUNCTION] - callback for when step is first displayed}

**onCTA** [FUNCTION] - callback for the optional call-to-action button}

Setting tour options
--------------------
Tour options can be specified either through the tour JSON object, or through a call to hopscotch.configure().

**bubbleWidth** [NUMBER] - Default bubble width. *Default*: 280.

**bubblePadding** [NUMBER] - Default bubble padding. *Default*: 15.

**smoothScroll** [BOOLEAN] - Should the page scroll smoothly to the next step? *Default*: true.

**scrollDuration** [NUMBER] - Duration of page scroll in milliseconds. Only relevant when smoothScroll is set to true. *Default*: 1000.

**scrollTopMargin** [NUMBER] - When the page scrolls, how much space should there be between the bubble/targetElement and the top of the viewport? *Default*: 200.

**showCloseButton** [BOOLEAN] - Should the tour bubble show a close (X) button? *Default*: true.

**showPrevButton** [BOOLEAN] - Should the bubble have the Previous button? *Default*: false.

**showNextButton** [BOOLEAN] - Should the bubble have the Next button? *Default*: true.

**arrowWidth** [NUMBER] - Default arrow width. (space between the bubble and the targetEl) Used for bubble position calculation. This option is provided for the case where the developer wants to use custom CSS to adjust the size of the arrow. *Default*: 20.

**skipIfNoElement** [BOOLEAN] - If a specified target element is not found, should we skip to the next step? *Default*: true.

**nextOnTargetClick** [BOOLEAN] - Should we advance to the next step when the user clicks on the target? *Default*: false.

**onNext** [FUNCTION] - Invoked after every click on a "Next" button.

**onPrev** [FUNCTION] - Invoked after every click on a "Prev" button.

**onStart** [FUNCTION] - Invoked when the tour is started.

**onEnd** [FUNCTION] - Invoked when the tour ends.

**onClose** [FUNCTION] - Invoked when the user closes the tour before finishing.

**onError** [FUNCTION] - Invoked when the specified target element doesn't exist on the page.

**i18n** [OBJECT] - For i18n purposes. Allows you to change the text of button labels and step numbers.

**i18n.nextBtn** [STRING] - Label for next button

**i18n.prevBtn** [STRING] - Label for prev button

**i18n.doneBtn** [STRING] - Label for done button

**i18n.skipBtn** [STRING] - Label for skip button

**i18n.closeTooltip** [STRING] - Text for close button tooltip

**i18n.stepNums** [ARRAY<STRING>] - Provide a list of strings to be shown as the step number, based on index of array. Unicode characters are supported. (e.g., ['&#x4e00;', '&#x4e8c;', '&#x4e09;']) If there are more steps than provided numbers, Arabic numerals ('4', '5', '6', etc.) will be used as default.

Running, Managing, and Ending a Tour
====================================

The Hopscotch framework comes with a simple set of API calls:

* `hopscotch.startTour(tour, stepNum)` - Actually starts the tour. Optional stepNum argument specifies what step to start at.
* `hopscotch.showStep(idx)` - Skips to a given step in the tour
* `hopscotch.prevStep()` - Goes back one step in the tour
* `hopscotch.nextStep()` - Goes forward one step in the tour
* `hopscotch.endTour([clearCookie])` - Ends the current tour. If clearCookie is set to false, the tour state is preserved. Otherwise, if clearCookie is set to true or is not provided, the tour state is cleared.
* `hopscotch.configure(options)` - Sets options for running the tour. Note: if this method is called after loading a tour, the options specified will override the options defined in the tour. See above section "Setting tour options" for a list of configuration options.

Defining callbacks
==================

Hopscotch has several events to which you can assign callbacks. These events include *start*, *end*, *next*, *prev*, *show*, *close*, *error*. For the *next*, *prev*, and *show* events, you can assign callbacks within step definitions as well as in the tour itself.

There are two ways to define event callbacks:

Function literals
-----------------
If you are specifying your tour as an object literal in Javascript, you can provide a function literal as the value of your callback. This would look like the following:

    var tour = {
      id: 'myTour',
      steps: [
        {
          target: 'firststep',
          orientation: 'bottom',
          title: 'My First Step',
          onNext: function() {
            $('#firststep').hide();
          }
        }
      ],
      onStart: function() {
        $('#article').addClass('selected');
      }
    };

Helpers
-------
In some situations, you may want to specify your tour in JSON. This may be because you are dynamically creating a tour on the server side. Since functions are not valid JSON, specifying a function literal will not work. Instead, you can use Hopscotch helpers to specify your callback. Using helpers will look something like the following.

First register the helper with Hopscotch.

    hopscotch.registerHelper('selectArticle', function() {
      $('#article').addClass('selected');
    });

An example with an argument passed in:

    hopscotch.registerHelper('fillText', function(textFieldId, textStr) {
      document.getElementById(textFieldId).value = textStr;
    });

Then, when you define your tour, you specify the callback as an array of the following form: `[helperId, arg, arg, ...]`. For example:

    {
      id: "myTour",
      steps: [
        {
          target: "firststep",
          orientation: "bottom",
          title: "My First Step",
          onNext: ["fillText", "searchField", "Example search query"]
        }
      ],
      onStart: ["selectArticle"]
    }

In the above example, since the onStart callback has no arguments, it could be defined as a simple string "selectArticle" instead of being wrapped in a one-element array.

To specify several helpers for a certain event:

    {
      id: "myTour",
      steps: [
        ...
      ],
      onStart: [["fillText", "searchField", "Example search query"], "selectArticle"]
    }

Callbacks will be invoked in the order that they are specified.

To remove a helper, simply call `hopscotch.unregisterHelper('myHelperId')`.


CHANGELOG.md
## Change Log

### v0.3.0 (2017/06/02)
 * [#337](https://github.com/linkedin/hopscotch/pull/337) Refactor build to generate UMD and AMD artifacts

### v0.2.8 (2017/05/04)
 * [#323](https://github.com/linkedin/hopscotch/pull/323) Fix the rest of the deprecated gradient warnings

### v0.2.7 (2017/02/24)
 * [#312](https://github.com/linkedin/hopscotch/pull/312) Generate templates without `with` keyword (now strict compatible)
 * [#307](https://github.com/linkedin/hopscotch/pull/307) Avoid "Gradient has outdated syntax" error
 * [#320](https://github.com/linkedin/hopscotch/pull/320) [Chore] Refactor notice; Bump year

### v0.2.6 (2016/07/01)
 * [#287](https://github.com/linkedin/hopscotch/pull/287) Add new items to .gitignore (ported over from gh72).
 * [#275](https://github.com/linkedin/hopscotch/pull/275) Update jasmine version in package.json
 * [#273](https://github.com/linkedin/hopscotch/pull/273) Fix unlisten method (wrong callback equals)
 * [#253](https://github.com/linkedin/hopscotch/pull/253) Improve documentation of the target attribute of step elements.
 * [#280](https://github.com/linkedin/hopscotch/pull/280) Miss a variable declaration
 * [#291](https://github.com/linkedin/hopscotch/pull/291) Add error callback for showStep (Issue #274)

### v0.2.5 (2015/07/20)
 * [#188](https://github.com/linkedin/hopscotch/pull/188) [GH-188] No event called after tour is finished
 * [#193](https://github.com/linkedin/hopscotch/pull/193) [GH-190] Close button should be a button
 * [#187](https://github.com/linkedin/hopscotch/pull/187) [GH-24] Recover when tour bubble DOM element is destroyed
 * [#183](https://github.com/linkedin/hopscotch/pull/183) Remove moot `version` property from bower.json
 * [#181](https://github.com/linkedin/hopscotch/pull/181) Remove all styling from hopscotch.less
 * [#182](https://github.com/linkedin/hopscotch/pull/182) [GH-175] totalSteps should be i18n-ified
 * [#178](https://github.com/linkedin/hopscotch/pull/178) [GH-177] Require existing target element for stand-alone callouts
 * [#176](https://github.com/linkedin/hopscotch/pull/176) [GH-172] documentIsReady is when document.readyState is complete

### v0.2.4 (2015/04/28)
 * [#168](https://github.com/linkedin/hopscotch/pull/168) Gracefully handle steps without valid target
 * [#167](https://github.com/linkedin/hopscotch/pull/167) Hide previous button on first step, even if steps are skipped.
 * [#153](https://github.com/linkedin/hopscotch/pull/153) Tour ID as part of classes
 * [#151](https://github.com/linkedin/hopscotch/pull/151) Add support for AMD and CommonJS
 * [#146](https://github.com/linkedin/hopscotch/pull/146) Check tour & callout IDs for invalid characters
 * [#142](https://github.com/linkedin/hopscotch/pull/142) Simple bower config file added.

### v0.2.3 (2014/12/10)

 * [#139](https://github.com/linkedin/hopscotch/pull/139) RefreshBubblePosition updates all known callouts
 * [#136](https://github.com/linkedin/hopscotch/pull/136) Ignore skipped steps in numbering
 * [#137](https://github.com/linkedin/hopscotch/pull/137) NextOnTarget Click event not always removed
 * [#133](https://github.com/linkedin/hopscotch/pull/133) Add right-to-left support to hopscotch
 * [#128](https://github.com/linkedin/hopscotch/pull/128) Trailing comma in opts.i18n object
 * [#126](https://github.com/linkedin/hopscotch/pull/126) Revert defuault z-index to empty vs auto
 * [#102](https://github.com/linkedin/hopscotch/pull/102) Callbacks on callouts do not work


### v0.2.2 (2014/06/26)

 * [#91](https://github.com/linkedin/hopscotch/pull/91) Don't cache step targets - support for single page apps
 * [#96](https://github.com/linkedin/hopscotch/pull/96) box-sizing: content-box - Issues with Bootstrap 3
 * [#90](https://github.com/linkedin/hopscotch/pull/90) Prefer jQuery and Sizzle when searching for a target
 * [#94](https://github.com/linkedin/hopscotch/pull/94) Catch exceptions caused by illegal targets in document.QuerySelector

### v0.2.1 (2014/06/05)

 * [#82](https://github.com/linkedin/hopscotch/pull/82) IE8 Button Bindings Fix
 * [#87](https://github.com/linkedin/hopscotch/pull/87) Fix blank z-index for value 0
 * [#89](https://github.com/linkedin/hopscotch/pull/89) Fix sessionStorage exceeds quota error in Safari

### v0.2.0 (2014/03/24)

 * [#70](https://github.com/linkedin/hopscotch/pull/70) Rendering tour & callout bubbles using compiled templates (@zimmi88)
 * [#62](https://github.com/linkedin/hopscotch/pull/62) Add method to redraw the bubble at a new position (@cedrics)
 * [#64](https://github.com/linkedin/hopscotch/pull/64) Add changelog (@kate2753)
 * [#61](https://github.com/linkedin/hopscotch/pull/61) Reorganize repo and automate release process, travis integration


### v0.1.3 (2014/03/04)

 * [aaca52b](https://github.com/linkedin/hopscotch/commit/aaca52bc2fcb254f65a5b0e3173635de7c9bb8d6) Update README: add explanation for the getCurrTarget() (@gnatok)
 * [d5894b2](https://github.com/linkedin/hopscotch/commit/d5894b24150597d38b6225e0b6040887d531a66c) Move window.sessionStorage into try catch block. (@nanek)
 * [35a2b2d](https://github.com/linkedin/hopscotch/commit/35a2b2d4f334cbb191899b860d4be80d420d4dcc) detect when a onNext/onPrev callback updates tour state (@gkoo)
 * [4c025eb](https://github.com/linkedin/hopscotch/commit/4c025ebf16023ce1e74dbfd34cf89cc9bbbd8ebb) a little cleanup from the previous commit (@gkoo)
 * [25140fb](https://github.com/linkedin/hopscotch/commit/25140fb3d1488930e0698093fb4e2e675b5108b8) Changed logic to check for typeof currStepNum. Forcing tour to start at step 0 was not working. (@leedavidr)
 * [bc2af92](https://github.com/linkedin/hopscotch/commit/bc2af92d025823decb3be3a3c32670c50cb45592) makes onCTA use callback helpers (@yeah)
 * [7791d6f](https://github.com/linkedin/hopscotch/commit/7791d6f346e1b2849894878aba48725199fb709a) Small doc update for query selectors. (@marc-hughes)

demo/js/debug.js
/* globals hopscotch:false */
var i,
    len,
    optionName,
    checkbox,
    fn,
    booleanControls = [
      'animate',
      'smoothScroll',
      'showCloseButton',
      'showPrevButton',
      'showNextButton'
    ],

addClickListener = function(el, fn) {
  if (el.addEventListener) {
    el.addEventListener('click', fn, false);
  }
  else {
    el.attachEvent('onclick', fn);
  }
},

attachListenerForOption = function(checkbox, optionName) {
  var fn = function() {
    var options = {};
    options[optionName] = checkbox.checked;
    hopscotch.configure(options);
  };
  if (checkbox.addEventListener) {
    checkbox.addEventListener('change', fn, false);
  }
  else {
    checkbox.attachEvent('onchange', fn);
  }
};

// BOOLEAN CONTROLS
for (i = 0, len = booleanControls.length; i < len; i++) {
  optionName = booleanControls[i];
  checkbox = document.getElementById(optionName + 'Check');
  attachListenerForOption(checkbox, optionName);
}

// SHOW ASIAN STEP NUMS
(function() {
  var checkbox = document.getElementById('showAsianStepNumsCheck'),
  fn = function() {
    console.log('changing steps');
    var asianSteps = [
          '&#x4e00;',
          '&#x4e8c;',
          '&#x4e09;',
          '&#x56db;',
          '&#x4e94;',
          '&#x516d;'
        ],
        steps = checkbox.checked ? asianSteps : null;

    hopscotch.configure({ i18n: { stepNums: steps } });
  };

  if (checkbox.addEventListener) {
    checkbox.addEventListener('change', fn, false);
  }
  else {
    checkbox.attachEvent('onchange', fn);
  }
}());

(function() {
  // GO TO STEP NUM
  var gotobtn  = document.getElementById('gotobtn');
  addClickListener(gotobtn, function() {
    var gototext = document.getElementById('gototext');
    hopscotch.showStep(parseInt(gototext.value, 10)-1);
  });

  // ==========
  // START TOUR
  // ==========
  addClickListener(document.getElementById('startBtn'), function() {
    hopscotch.startTour(tour);
  });

  // ========
  // END TOUR
  // ========
  addClickListener(document.getElementById('endBtn'), function() {
    hopscotch.endTour();
  });

  // ===========
  // ADD CALLOUT
  // ===========
  addClickListener(document.getElementById('addCallout'), function() {
    var mgr = hopscotch.getCalloutManager();
    mgr.createCallout({
      id: 'mycallout',
      title: 'Title',
      content: 'Content',
      target: 'google',
      orientation: 'top',
      showCTAButton: true,
      ctaLabel: 'hello',
      onCTA: function() { alert('test'); },
      showNumber: false
    });
  });

  // ==============
  // REMOVE CALLOUT
  // ==============
  addClickListener(document.getElementById('rmCallout'), function() {
    var mgr = hopscotch.getCalloutManager();
    mgr.removeCallout('mycallout');
  });

  // ===========
  // START TOUR2
  // ===========
  //addClickListener(document.getElementById('startBtn2'), function() {
    //hopscotch.startTour(tour2);
  //});
}());

var tour2 = {
  id: 'hello-hopscotch2',
  steps: [
    {
      title: 'Debug controls',
      content: 'Here are the debug controls. They\'re pretty self-explanatory: start a tour, end a tour, clear the tour cookie (if you want the tour to forget what step you are on).',
      targetId: 'debug',
      orientation: 'left',
      width: 320,
      height: 480,
      onPrev: function() {
        document.getElementById('pageTitle').style.color = '#000';
      }
    },
    {
      title: 'Python decorator',
      content: 'Whoa, did you notice that the page just scrolled? If you didn\'t, you aren\'t very observant. (Or you have a very tall monitor)',
      targetId: 'google',
      orientation: 'top',
      xOffset: 200
    }
  ],
  animate: false,
  //smoothScroll: false,
  //showCloseButton: false,
  //scrollTopMargin: 200,
  showPrevButton: true,
  showNextButton: true,
  arrowWidth: 20,
  scrollDuration: 500,
  cookieName: 'li_hs',
  onNext: function(tourId, idx) {
    var newLi,
        list = document.getElementById('my-list');
    if (list) {
      newLi = document.createElement('li');
      newLi.innerHTML = 'going from step ' + (idx + 1) + ' to step ' + (idx+2) + '. (can use this callback for tracking)';
      list.appendChild(newLi);
    }
  },
  onPrev: function(tourId, idx) {
    console.log('tour2 prev ' + tourId + ' ' + idx);
  },
  onStart: function(tourId) {
    console.log('tour2 start ' + tourId);
  },
  onEnd: function(tourId) {
    console.log('tour2 end ' + tourId);
  },
  i18n: {
    //nextBtn: 'Forward',
    //prevBtn: 'Backward',
    stepNums: [
      '&#x4e00;',
      '&#x4e8c;',
      '&#x4e09;',
      '&#x56db;',
      '&#x4e94;',
      '&#x516d;'
    ]
  }
};
